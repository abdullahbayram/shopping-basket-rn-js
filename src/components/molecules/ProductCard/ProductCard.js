import * as React from 'react';
import { Button, Text } from 'react-native-paper';
import PropTypes from 'prop-types';
import { StyleSheet } from 'react-native';
import Card from '../../atoms/Card/Card';

const renderRightContent = (price) => <Text variant="titleSmall">${price}</Text>;

const ProductCard = ({ buttonTitle, subtitle, title, price, onButtonPress }) => (
  <Card style={styles.container}>
    <Card.Title style={styles.title} title={title} right={() => renderRightContent(price)} subtitle={subtitle} />
    <Card.Actions>
      <Button onPress={onButtonPress}>{buttonTitle}</Button>
    </Card.Actions>
  </Card>
);

export default ProductCard;

ProductCard.propTypes = {
  buttonTitle: PropTypes.string.isRequired,
  subtitle: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  price: PropTypes.string.isRequired,
  onButtonPress: PropTypes.func.isRequired,
};

const styles = StyleSheet.create({
  container: {
    marginVertical: 10,
  },
  title: {
    flexDirection: 'row',
    paddingHorizontal: 16,
  },
});
