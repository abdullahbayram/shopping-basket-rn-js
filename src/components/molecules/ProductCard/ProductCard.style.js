import { StyleSheet } from 'react-native';
import { typography } from '@constants/theme';

export default (colors, spacing) =>
  StyleSheet.create({
    container: {
      marginVertical: spacing.sm,
      borderRadius: spacing.sm,
      flex: 1,
      backgroundColor: colors.cardBackground,
      shadowColor: colors.shadowColor,
      shadowOffset: { width: 0, height: 2 },
      shadowOpacity: 0.2,
      shadowRadius: 4,
      elevation: 4,
    },
    title: {
      flexDirection: 'row',
      paddingVertical: 0,
      paddingBottom: 0,
      marginVertical: 0,
    },
    rating: {
      fontSize: typography.sm,
      color: colors.ratingStar,
      marginBottom: spacing.xs,
    },
    feedbackAndPrice: {
      paddingHorizontal: spacing?.md,
    },
    price: {
      fontSize: typography.default,
      fontWeight: typography.bold,
      color: colors.textTertiary,
    },
    button: {
      borderRadius: 50,
      flex: 1,
      backgroundColor: colors.secondary,
    },
    disabledButton: {
      backgroundColor: colors.disabledButton,
      borderColor: colors.disabledButtonBorder,
    },
    buttonOrHelperTextContainer: {
      paddingHorizontal: spacing.special,
      flexDirection: 'row',
      height: 75,
      alignItems: 'center',
      justifyContent: 'center',
    },
  });
