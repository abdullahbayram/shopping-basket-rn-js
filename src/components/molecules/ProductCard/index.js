import React from 'react';
import PropTypes from 'prop-types';
import { View } from 'react-native';
import { Card, Button, Text, HelperText } from '../../atoms';
import styles from './ProductCard.style';

const leftCardMargin = { marginRight: 5 };
const rightCardMargin = { marginLeft: 5 };

const generateCardMargin = (index) => (index % 2 === 0 ? leftCardMargin : rightCardMargin);

const ProductCard = ({ product, onButtonPress, isMaxQuantityPerProductReached, index = 0 }) => {
  const { title, price, rating, image } = product;
  const maxStars = 5;
  const filledStars = Math.round((rating.rate / maxStars) * maxStars);
  const stars = '★'.repeat(filledStars) + '☆'.repeat(maxStars - filledStars);

  return (
    <Card style={[styles.container, generateCardMargin(index)]}>
      <Card.Cover source={{ uri: image }} />
      <Card.Title style={styles.title} title={title} subtitle={stars} subtitleStyle={styles.rating} />
      <View style={styles.feedbackAndPrice}>
        <Text style={styles.price}>€{price.toFixed(2)}</Text>
      </View>

      <View style={styles.buttonOrHelperTextContainer}>
        {!isMaxQuantityPerProductReached ? (
          <Button onPress={onButtonPress} style={styles.button} disabled={isMaxQuantityPerProductReached}>
            Add to basket
          </Button>
        ) : (
          <HelperText type="error" visible>
            You reached the max quantity per product!
          </HelperText>
        )}
      </View>
    </Card>
  );
};

ProductCard.whyDidYouRender = true;

export default ProductCard;

ProductCard.propTypes = {
  product: PropTypes.shape({
    id: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
    title: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    description: PropTypes.string.isRequired,
    category: PropTypes.string.isRequired,
    image: PropTypes.string.isRequired,
    rating: PropTypes.shape({
      rate: PropTypes.number.isRequired,
      count: PropTypes.number.isRequired,
    }).isRequired,
  }).isRequired,
  onButtonPress: PropTypes.func.isRequired,
  isMaxQuantityPerProductReached: PropTypes.bool.isRequired,
  index: PropTypes.number.isRequired,
};
