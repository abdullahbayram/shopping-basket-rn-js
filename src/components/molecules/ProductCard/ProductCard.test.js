import React from 'react';
import { fireEvent, screen } from '@testing-library/react-native';
import { renderInThemeProvider } from '@testUtils/renderInThemeProvider';
import ProductCard from '.';

describe('<ProductCard />', () => {
  const mockOnButtonPress = jest.fn();
  const mockProduct = {
    id: 1,
    title: 'Test Product',
    price: 19.99,
    description: 'A wonderful test product',
    category: 'Test Category',
    image: 'https://example.com/image.jpg',
    rating: {
      rate: 4.5,
      count: 100,
    },
  };

  const renderProductCard = (props) => {
    return renderInThemeProvider(
      <ProductCard
        product={mockProduct}
        onButtonPress={mockOnButtonPress}
        isMaxQuantityPerProductReached={false}
        index={0}
        {...props}
      />,
    );
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly with product details', () => {
    renderProductCard();

    expect(screen.getByText('Test Product')).toBeTruthy();
    expect(screen.getByText('€19.99')).toBeTruthy();
    expect(screen.getByText('★★★★★')).toBeTruthy();
  });

  it('matches the snapshot', () => {
    const { toJSON } = renderProductCard();
    expect(toJSON()).toMatchSnapshot();
  });

  it('calls onButtonPress when the Add to basket button is pressed', () => {
    renderProductCard();

    const addButton = screen.getByText('Add to basket');
    fireEvent.press(addButton);

    expect(mockOnButtonPress).toHaveBeenCalled();
  });

  it('displays the helper text when max quantity per product is reached', () => {
    renderProductCard({ isMaxQuantityPerProductReached: true });

    expect(screen.getByText('You reached the max quantity per product!')).toBeTruthy();
    const addButton = screen.queryByText('Add to basket');
    expect(addButton).toBeNull();
  });

  it('renders left margin for even index', () => {
    renderProductCard({ index: 0 });
    const card = screen.getByTestId('product-card-container-outer-layer');

    expect(card.props.style).toEqual(expect.objectContaining({ marginRight: 5 }));
  });

  it('renders right margin for odd index', () => {
    renderProductCard({ index: 1 });
    const card = screen.getByTestId('product-card-container-outer-layer');

    expect(card.props.style).toEqual(expect.objectContaining({ marginLeft: 5 }));
  });
});
