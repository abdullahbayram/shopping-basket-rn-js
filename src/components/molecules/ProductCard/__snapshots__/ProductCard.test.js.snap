// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCard /> matches the snapshot 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "#F0F0F0",
      "borderRadius": 10,
      "flex": 1,
      "marginRight": 5,
      "marginVertical": 10,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 1,
        "width": 0,
      },
      "shadowOpacity": 0.15,
      "shadowRadius": 3,
    }
  }
  testID="product-card-container-outer-layer"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#F0F0F0",
        "borderRadius": 10,
        "elevation": 4,
        "flex": undefined,
        "shadowColor": "rgba(0, 0, 0, 0.3)",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.2,
        "shadowRadius": 4,
      }
    }
    testID="product-card-container"
  >
    <View
      style={
        [
          {
            "flexShrink": 1,
          },
          undefined,
        ]
      }
      testID="product-card"
    >
      <View
        style={
          [
            {
              "backgroundColor": "#eeeeee",
              "height": 195,
              "overflow": "hidden",
            },
            {
              "borderRadius": 6,
            },
            undefined,
          ]
        }
      >
        <Image
          accessibilityIgnoresInvertColors={true}
          borderRadiusStyles={
            {
              "borderRadius": 10,
            }
          }
          siblings={
            [
              "Card.Cover",
              "Card.Title",
              "View",
              "View",
            ]
          }
          source={
            {
              "uri": "https://example.com/image.jpg",
            }
          }
          style={
            [
              {
                "flex": 1,
                "height": undefined,
                "justifyContent": "flex-end",
                "padding": 16,
                "width": undefined,
              },
              {
                "borderRadius": 6,
              },
            ]
          }
        />
      </View>
      <View
        style={
          [
            {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingLeft": 16,
            },
            {
              "minHeight": 72,
            },
            {
              "flexDirection": "row",
              "marginVertical": 0,
              "paddingBottom": 0,
              "paddingVertical": 0,
            },
          ]
        }
      >
        <View
          style={
            [
              {
                "flex": 1,
                "flexDirection": "column",
                "justifyContent": "center",
              },
            ]
          }
        >
          <Text
            numberOfLines={1}
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "letterSpacing": 0.15,
                    "lineHeight": 24,
                  },
                  [
                    {
                      "minHeight": 30,
                      "paddingRight": 16,
                    },
                    {
                      "marginBottom": 0,
                    },
                    undefined,
                  ],
                ],
              ]
            }
          >
            Test Product
          </Text>
          <Text
            numberOfLines={1}
            style={
              [
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "writingDirection": "ltr",
                },
                [
                  {
                    "fontFamily": "System",
                    "fontSize": 14,
                    "fontWeight": "400",
                    "letterSpacing": 0.25,
                    "lineHeight": 20,
                  },
                  [
                    {
                      "marginVertical": 0,
                      "minHeight": 20,
                      "paddingRight": 16,
                    },
                    {
                      "color": "#FFA500",
                      "fontSize": 12,
                      "marginBottom": 4,
                    },
                  ],
                ],
              ]
            }
          >
            ★★★★★
          </Text>
        </View>
        <View />
      </View>
      <View
        borderRadiusStyles={
          {
            "borderRadius": 10,
          }
        }
        index={2}
        siblings={
          [
            "Card.Cover",
            "Card.Title",
            "View",
            "View",
          ]
        }
        style={
          {
            "paddingHorizontal": 16,
          }
        }
        total={4}
      >
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              [
                {
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "400",
                  "letterSpacing": 0.25,
                  "lineHeight": 20,
                },
                [
                  {
                    "color": "#34300F",
                  },
                  {
                    "color": "#000",
                    "fontSize": 16,
                    "fontWeight": "700",
                  },
                ],
              ],
            ]
          }
        >
          €
          19.99
        </Text>
      </View>
      <View
        borderRadiusStyles={
          {
            "borderRadius": 10,
          }
        }
        index={3}
        siblings={
          [
            "Card.Cover",
            "Card.Title",
            "View",
            "View",
          ]
        }
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 75,
            "justifyContent": "center",
            "paddingHorizontal": 7,
          }
        }
        total={4}
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "#FED034",
              "borderRadius": 50,
              "flex": 1,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
            }
          }
          testID="button-container-outer-layer"
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "#FED034",
                "borderColor": "rgba(121, 116, 126, 1)",
                "borderRadius": 50,
                "borderStyle": "solid",
                "borderWidth": 1,
                "flex": undefined,
                "minWidth": 64,
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
              }
            }
            testID="button-container"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "overflow": "hidden",
                  },
                  {
                    "borderRadius": 50,
                  },
                ]
              }
              testID="button"
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    [
                      {
                        "textAlign": "left",
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "writingDirection": "ltr",
                      },
                      [
                        {
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        [
                          {
                            "marginHorizontal": 16,
                            "marginVertical": 9,
                            "textAlign": "center",
                          },
                          false,
                          {
                            "marginHorizontal": 24,
                            "marginVertical": 10,
                          },
                          undefined,
                          false,
                          {
                            "color": "#34300F",
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "500",
                            "letterSpacing": 0.1,
                            "lineHeight": 20,
                          },
                          undefined,
                        ],
                      ],
                    ]
                  }
                  testID="button-text"
                >
                  Add to basket
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
